### ISSUE TITLE
Separate modification and saving of session data

### ISSUE THREAD

**Issue by alexedwards** [alexedwards/scs#36]
See discussion in https://github.com/alexedwards/scs/issues/23

This could be addressed by implementing Middleware which uses a custom ResponseWriter to intercept HTTP Responses, save the session data to the store and set the session cookie  - similar to v0.1 of SCS (https://github.com/alexedwards/scs/blob/v0.1.0/session/manager.go)

Downside is that this causes problems with frameworks/applications which also try to call WriteHeader. See https://github.com/alexedwards/scs/issues/15

A workaround might be to set the session cookie header each time the session is modified (or even loaded) and only handle saving to the store in the middleware. I think that would work for most stores, except the CookieStore, which *needs* the response headers to *not* have already been written.

**Comment by jpfluger** [comment]
Ultimately scs is one key ingredient to a greater solution. Since the greater solution is often based on a framework (with its own custom http handlers), it seems that fixing this issue is best left to middleware. I do something similar in my middleware, where I use a `defer func` in the middleware to run `PutObject`. This design saves on multiple writes but only if data is stored as a struct and my working copy is reachable outside scs.data. In my situation RAM is cheap but network writes more expensive. 

I like the idea of [AutoSave](https://github.com/alexedwards/scs/issues/23#issuecomment-346958129) but not certain if it structurally makes sense to do both. (referring to dhui's [comments](https://github.com/alexedwards/scs/issues/23#issuecomment-357360643)) But maybe it's possible. In looking at the [put](https://github.com/alexedwards/scs/blob/2b6a87b9c81e34a20d7e1c8a2be729c67e252566/session.go#L742) code, you could try....

```golang
	if autoSave {
		saveCounter++
		return nil
	}

	return s.write(w)
}
```

Then a different function is maybe `func ShouldSave() bool` or `func SaveCounter() int`. If either is true, then the middleware needs to call a new `func Save() error`.

**Comment by alexedwards** [comment]
This issue was fixed as part of the v2.0.0 release.
